src/metabase/upload.clj

```diff
@@ -483,3 +489,133 @@
+  (let [table-cols           (t2/select :model/Field :table_id (:id table))
+        table-cols-by-normed (-> (into {}
+                                       (map (fn [col]
+                                              [(normalize-column-name (:name col)) col]))
+                                       table-cols)
+                                 (dissoc "_mb_row_id"))
+        normalized-header+header (into []
+                                       (map (fn [col]
+                                              [(normalize-column-name col) col]))
+                                       header)
+        normed first
+        normalized-header (map normed normalized-header+header)
+        normalized-table-cols (map normed table-cols-by-normed)
+        ;; check for duplicates
+        _ (when (some #(< 1 %) (vals (frequencies normalized-header)))
+            (throw (ex-info (tru "The CSV file contains duplicate column names.")
+                            {:status-code 422})))
+        [extra missing _both] (data/diff (set normalized-header)
+                                         (set normalized-table-cols))]
+
+    (if (and (nil? extra) (nil? missing))
+      {:parse-rows (fn [rows]
+                     (let [rows' rows ;; TODO: remove auto-pk columns here
+                           col-upload-types (map (comp base-type->upload-type :base_type table-cols-by-normed) normalized-header)]
+                       (parse-rows* col-upload-types rows')))
+       :table-col-names (map (comp :name table-cols-by-normed) normalized-header)}
+      (throw (ex-info (tru "The schema of the CSV file does not match the schema of the table.")
+                      {:status-code 422})))))
+
+(defn- append-csv!*
+  [database table file]
+  (with-open [reader (bom/bom-reader file)]
+    (let [[header & rows] (csv/read-csv reader)
+          {:keys [parse-rows table-col-names]} (check-schema table header)
+          schema+table-name (if (str/blank? (:schema table))
+                              (:name table)
+                              (str (:schema table) "." (:name table)))]
+      (try
+        (let [parsed-rows (parse-rows rows)]
+          (driver/insert-into! (driver.u/database->driver database)
+                               (:id database)
+                               schema+table-name
+                               table-col-names
+                               parsed-rows)
+          {:row-count (count parsed-rows)})
+        (catch Throwable e
+          ;; TODO: insert rows failure handling
+          (throw (ex-info (ex-message e) {:status-code 400})))))))
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                                 public interface                                               |
+;;; +----------------------------------------------------------------------------------------------------------------+
```

Comment:
this doesn't align with other blocks, for one this contains more than "public" interface and also the template is different.